proto/src/proto.rs

```diff
@@ -372,7 +372,34 @@ pub enum LdapDerefAliases {
 pub struct LdapSubstringFilter {
     pub initial: Option<String>,
     pub any: Vec<String>,
-    pub final_: Option<String>,
+    pub final_: Option<String>, //escape final keyword
+}
+
+impl From<&str> for LdapSubstringFilter {
+    fn from(value: &str) -> Self {
+        let mut filter = LdapSubstringFilter {
+            initial: None,
+            any: Vec::new(),
+            final_: None,
+        };
+
+        let value_count = value.split('*').count();
+        let last_char = value.chars().last();
+        let first_char = value.chars().next();
+        value.split('*').enumerate().for_each(|(idx, v)| {
+            if idx == 0 && first_char != Some('*') {
+                filter.initial = Some(v.to_string())
+            } else if idx == value_count - 1 && last_char != Some('*') {
+                filter.final_ = Some(v.to_string())
+            } else if v.is_empty() {
+                // pass
+            } else {
+                filter.any.push(v.to_string())
+            }
+        });
```

Comment:
I think it suffers the exact same code flow problem as the match version, but with ifs instead. Let me elaborate. When reading this code it’s not immediately obvious which condition is relevant or not at which iteration. In this case, the first condition is only relevant for the first iteration, and the second one is only relevant for the last iteration while the other one is relevant for all remaining intermediate iterations (but not the first nor the last). This is a [clever way](https://guifroes.com/clever-code-is-bad/) to write all the conditions in a single loop, but it also checks for conditions that are not (yet/anymore) relevant for each intermediate iteration. [This suggestion](https://github.com/kanidm/ldap3/pull/40#discussion_r1412533940) was more about splitting the operation into three distinct steps clearly executed sequentially. Well, this is a small detail, but I hope this gives some insight to my initial suggestion. Of course, it’s not a blocker for merging the PR either. :slightly_smiling_face:

Reply:
I think I agree here, because while this is clever, it is also dense. 

There may however be a simpler way. str split has a pretty simple property that if you have "*a*b*" then you get ["", "a", "b", "" ]

This is because it splits in a way where if nothing is to the left/right of the char, then you get an empty string.

Split is also a *double ended* iterator. This lets you pull the initial/end out first.

This would allow you to do:

```
let split_iter = value.split('*')
if let Some(start) = split_iter.next() {
    if !start.is_empty() {
        filter.initial = Some(start.to_string());
    }
}

if let Some(end) = split_iter.next_back() {
    if !end.is_empty() {
        filter.end = Some(end.to_string());
    }
}

split_iter.for_each(|v| {
    filter.any.push(v.to_string())
});
```

So provided I haven't overlooked anything wild, I think this should work for you.

Reply:
That's even better! 👍

Reply:
updated